File: usr/src/cmd/gui-install/src/help-dialog.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <glib/gstdio.h>
#include <gnome.h>

#include "installation-profile.h"
#include "interface-globals.h"
#include "window-graphics.h"
#include "help-dialog.h"

gboolean
show_file_in_textview(GtkWidget *textview,
						gchar *filename,
						gboolean bold,
						gboolean centered,
						gboolean dontprocessCR)
{
	FILE *fp;
	char ch, prevch;
	gboolean ret_val = TRUE;
	char str[MAXBUFFER];
	GtkTextBuffer *textBuffer = NULL;
	GtkTextTagTable *textTagTable = NULL;
	GtkTextTag *textTagBold = NULL;
	GtkTextTag *textTagUnderline = NULL;
	GtkTextTag *textTagCenter = NULL;
	int lineNum = 0, chNum = 0;
	GtkTextIter start, end;

	/* bold/centered paramaters apply only to the first line of input */

	if (!filename) {
		g_warning("Error : Must provide file name to read\n");
		return (FALSE);
	}

	if ((fp = fopen(filename, "r")) == NULL) {
		g_warning("Error : Failed to open file : %s\n", filename);
		return (FALSE);
	}

	textBuffer = gtk_text_view_get_buffer(GTK_TEXT_VIEW(textview));

	if ((!bold && !centered) && dontprocessCR) {
		/* Example would be install log, we just want to dump contents */
		/* into the textview without any formatting */
		while (fgets(str, MAXBUFFER, fp)) {
			gtk_text_buffer_insert_at_cursor(textBuffer, (const gchar *)str,
					strlen(str));
		}

	} else {
		if (bold || centered) {
			textTagTable = gtk_text_buffer_get_tag_table(textBuffer);

			if (textTagTable) {
				textTagBold = gtk_text_tag_table_lookup(textTagTable, "Bold");
				textTagUnderline =
					gtk_text_tag_table_lookup(textTagTable, "Underline");
				textTagCenter = gtk_text_tag_table_lookup(textTagTable,
															"Center");
			}

			if (!textTagBold) {
				textTagBold = gtk_text_buffer_create_tag(textBuffer, "Bold",
										"weight", PANGO_WEIGHT_BOLD, NULL);
			}
			if (!textTagUnderline) {
				textTagUnderline =
					gtk_text_buffer_create_tag(textBuffer, "Underline",
						"underline", PANGO_UNDERLINE_SINGLE, NULL);
			}
			if (!textTagCenter) {
				textTagCenter = gtk_text_buffer_create_tag(textBuffer,
										"Center", "justification",
										GTK_JUSTIFY_CENTER, NULL);
			}
		}

		lineNum = 0;
		chNum = 0;
		while ((ch = fgetc(fp)) != EOF) {
			if (ch == '\n') {
				lineNum++;
				if (dontprocessCR) {
					/* Just simply process the first line for BOLD/CENTER */
					if (lineNum == 1) {
						prevch = '\0';
						str[chNum] = '\n';
						str[chNum+1] = '\n';
						str[chNum+2] = '\0';
						gtk_text_buffer_insert_at_cursor(textBuffer,
											(const gchar *)str,
											strlen(str));
						chNum = 0;
						gtk_text_buffer_get_bounds(textBuffer, &start, &end);

						if (bold) {
							gtk_text_buffer_apply_tag_by_name(textBuffer,
													"Bold", &start, &end);
							gtk_text_buffer_apply_tag_by_name(textBuffer,
													"Underline", &start, &end);
						}
						if (centered) {
							gtk_text_buffer_apply_tag_by_name(textBuffer,
													"Center", &start, &end);
						}
					} else {
						prevch = ch;
						str[chNum] = ch;
						chNum++;

						if (chNum == (MAXBUFFER-2)) {
							str[chNum] = '\0';
							gtk_text_buffer_insert_at_cursor(textBuffer,
													(const gchar *)str,
													strlen(str));
							chNum = 0;
						}
					}
				} else {
					if (prevch == '\n' || lineNum == 1) {
						prevch = '\0';
						str[chNum] = '\n';
						str[chNum+1] = '\n';
						str[chNum+2] = '\0';
						gtk_text_buffer_insert_at_cursor(textBuffer,
											(const gchar *)str,
											strlen(str));
						chNum = 0;
						if (lineNum == 1 && (bold || centered)) {
							gtk_text_buffer_get_bounds(textBuffer, &start,
														&end);

							if (bold) {
								gtk_text_buffer_apply_tag_by_name(textBuffer,
													"Bold", &start, &end);
								gtk_text_buffer_apply_tag_by_name(textBuffer,
													"Underline", &start, &end);
							}
							if (centered) {
								gtk_text_buffer_apply_tag_by_name(textBuffer,
													"Center", &start, &end);
							}
						}
					} else {
						if (chNum > 0) {
							prevch = ch;
							str[chNum] = ' ';
							chNum++;
						}
					}
				}
			} else {
				prevch = ch;
				str[chNum] = ch;
				chNum++;

				if (chNum == (MAXBUFFER-2)) {
					str[chNum] = '\0';
					gtk_text_buffer_insert_at_cursor(textBuffer,
											(const gchar *)str,
											strlen(str));
					chNum = 0;
				}
			}
		}

		if (chNum > 0) {
			str[chNum] = '\0';
			gtk_text_buffer_insert_at_cursor(textBuffer, (const gchar *)str,
						strlen(str));
		}
	}

	return (ret_val);
}

gboolean
show_locale_file_in_textview(GtkWidget *textview,
				gchar *filename,
				gboolean bold,
				gboolean centered,
				gboolean dontprocessCR)
{
	gboolean ret_val = TRUE;
	FILE *fp = NULL;
	gint tfd = 0;
	gchar *tfilename = NULL;
	gchar *contents = NULL;
	gchar *contents_utf8 = NULL;
	gsize len = 0, len_utf8 = 0, bytes_read = 0;

	if (!filename) {
		g_warning("Error : Must provide file name to read\n");
		return (FALSE);
	}

	if ((fp = fopen(filename, "r")) == NULL) {
		g_warning("Error : Failed to open file : %s\n", filename);
		return (FALSE);
	}
	fclose(fp);

	if ((tfd = g_file_open_tmp("gui-install_localefile_XXXXXX",
					&tfilename, NULL)) == -1) {
		g_warning("Error : Failed to open temporary file.\n");
		return (FALSE);
	}
	close(tfd);

	if ((g_file_get_contents(filename, &contents, &len, NULL)) == FALSE) {
		g_warning("Error : Failed to get contents of file : %s\n", filename);
		g_unlink(tfilename);
		g_free(tfilename);
		return (FALSE);
	}

	if ((contents_utf8 = g_locale_to_utf8((const gchar *)contents, len,
			&bytes_read, &len_utf8, NULL)) == NULL) {
		g_warning("Error : Failed to convert contents of file to UTF-8 : %s\n",
			filename);
		g_free(contents);
		g_unlink(tfilename);
		g_free(tfilename);
		return (FALSE);
	}

	if ((g_file_set_contents((const gchar *)tfilename,
			(const gchar *)contents_utf8, len_utf8, NULL)) == FALSE) {
		g_warning("Error : Failed to write to tmp file : %s\n", tfilename);
		g_free(contents);
		g_free(contents_utf8);
		g_unlink(tfilename);
		g_free(tfilename);
		return (FALSE);
	}

	ret_val = show_file_in_textview(textview, tfilename,
						bold, centered, dontprocessCR);

	g_free(contents);
	g_free(contents_utf8);
	g_unlink(tfilename);
	g_free(tfilename);

	return (ret_val);
}

void
delete_textview_contents(GtkWidget *textview)
{
	GtkTextBuffer *textBuffer;
	textBuffer = gtk_text_view_get_buffer(GTK_TEXT_VIEW(textview));
	gtk_text_buffer_set_text(textBuffer, "", -1);
}

void
help_dialog_show(InstallScreen currScreen, gboolean bringToFront)
{
	gchar *tmpFileName = NULL;

	switch (currScreen) {
		case WELCOME_SCREEN :
			tmpFileName = MainWindow.TextFileLocations[HELP_WELCOME];
			break;
		case DISK_SCREEN :
			switch (InstallationProfile.installationtype) {
				case INSTALLATION_TYPE_INITIAL_INSTALL:
					tmpFileName =
								MainWindow.TextFileLocations[HELP_INSTALL_DISK];
					break;
				case INSTALLATION_TYPE_INPLACE_UPGRADE:
					tmpFileName =
								MainWindow.TextFileLocations[HELP_UPGRADE_DISK];
					break;
			}
			break;
		case TIMEZONE_SCREEN :
			tmpFileName = MainWindow.TextFileLocations[HELP_INSTALL_TIMEZONE];
			break;
		case LANGUAGE_SCREEN :
			tmpFileName = MainWindow.TextFileLocations[HELP_INSTALL_LANGUAGE];
			break;
		case USER_SCREEN :
			tmpFileName = MainWindow.TextFileLocations[HELP_INSTALL_USERS];
			break;
		case CONFIRMATION_SCREEN :
			switch (InstallationProfile.installationtype) {
				case INSTALLATION_TYPE_INITIAL_INSTALL:
					tmpFileName =
						MainWindow.TextFileLocations[HELP_INSTALL_CONFIRMATION];
					break;
				case INSTALLATION_TYPE_INPLACE_UPGRADE:
					tmpFileName =
						MainWindow.TextFileLocations[HELP_UPGRADE_CONFIRMATION];
					break;
			}
			break;
		case INSTALLATION_SCREEN :
			switch (InstallationProfile.installationtype) {
				case INSTALLATION_TYPE_INITIAL_INSTALL:
					tmpFileName =
							MainWindow.TextFileLocations[HELP_INSTALL_PROGRESS];
					break;
				case INSTALLATION_TYPE_INPLACE_UPGRADE:
					tmpFileName =
							MainWindow.TextFileLocations[HELP_UPGRADE_PROGRESS];
					break;
			}
			break;
		case FAILURE_SCREEN :
			switch (InstallationProfile.installationtype) {
				case INSTALLATION_TYPE_INITIAL_INSTALL:
					tmpFileName =
							MainWindow.TextFileLocations[HELP_INSTALL_FAILURE];
					break;
				case INSTALLATION_TYPE_INPLACE_UPGRADE:
					tmpFileName =
							MainWindow.TextFileLocations[HELP_UPGRADE_FAILURE];
					break;
			}
			break;
		case FINISH_SCREEN :
			tmpFileName = MainWindow.TextFileLocations[HELP_FINISH];
			break;
	}

	if (tmpFileName) {
		delete_textview_contents(MainWindow.helptextview);
		show_file_in_textview(MainWindow.helptextview, tmpFileName,
								TRUE, FALSE, TRUE);
		if (bringToFront) {
			window_graphics_dialog_set_properties(MainWindow.helpdialog);
			gtk_widget_show(MainWindow.helpdialog);
		}
	}
}

void
help_dialog_hide(GtkWidget *widget,
				gpointer *dialog)
{
	gtk_widget_hide(GTK_WIDGET(dialog));
}

void
help_dialog_delete_event(GtkWidget *widget,
				gpointer user_data)
{
	gtk_widget_hide(widget);
}

gchar *
help_generate_file_path(gchar *path, gchar *locale_id, gchar *filename)
{
	gchar *tmpFileName = NULL;
	FILE *fp;

	/* Test locale_id first */
	if (locale_id && strcmp("C", locale_id)) {
		if (filename) {
			tmpFileName = g_strdup_printf("%s/%s/%s", path, locale_id, filename);
		} else {
			tmpFileName = g_strdup_printf("%s/%s", path, locale_id);
		}

		if ((fp = fopen(tmpFileName, "r")) != NULL) {
			fclose(fp);
		} else {
			g_free(tmpFileName);
			tmpFileName = NULL;
		}
	}

	/* Did not find locale specific file so try for C Locale */
	if (!tmpFileName) {
		if (filename) {
			tmpFileName = g_strdup_printf("%s/C/%s", path, filename);
		} else {
			tmpFileName = g_strdup_printf("%s/C", path );
		}

		if ((fp = fopen(tmpFileName, "r")) != NULL) {
			fclose(fp);
		} else {
			g_free(tmpFileName);
			tmpFileName = NULL;
		}
	}

	/* Finally check if file exists without any locale information at all */
	/* e.g. install log etc */
	if (!tmpFileName) {
		if (filename) {
			tmpFileName = g_strdup_printf("%s/%s", path, filename);
		} else {
			tmpFileName = g_strdup_printf("%s", path);
		}

		if ((fp = fopen(tmpFileName, "r")) != NULL) {
			fclose(fp);
		} else {
			g_free(tmpFileName);
			tmpFileName = NULL;
		}
	}

	return (tmpFileName);
}

void
help_dialog_refresh(InstallScreen currScreen)
{
	/* Is Help Dialog displayed at the moment */
	if (GTK_WIDGET_VISIBLE(MainWindow.helpdialog)) {
		help_dialog_show(currScreen, FALSE);
	}
}

File: usr/src/cmd/gui-install/src/help_dialog.py
#!/usr/bin/python
#
# CDDL HEADER START
#
# The contents of this file are subject to the terms of the
# Common Development and Distribution License (the "License").
# You may not use this file except in compliance with the License.
#
# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
# or http://www.opensolaris.org/os/licensing.
# See the License for the specific language governing permissions
# and limitations under the License.
#
# When distributing Covered Code, include this CDDL HEADER in each
# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
# If applicable, add the following below this CDDL HEADER, with the
# fields enclosed by brackets "[]" replaced with your own identifying
# information: Portions Copyright [yyyy] [name of copyright owner]
#
# CDDL HEADER END
#
# Copyright (c) 2011, 2012, Oracle and/or its affiliates. All rights reserved.
#

'''
Help Dialog for GUI Install app
'''

import pygtk
pygtk.require('2.0')

import os

import g11nsvc
import gtk
import pango

from solaris_install.gui_install.gui_install_common import modal_dialog, \
    GLADE_DIR, GLADE_ERROR_MSG

# The HELPFILES list will need to be reordered once the UPGRADE panel(s)
# are completed.
HELPFILES = ['WELCOME_PANEL.txt',
             'INSTALL_DISK_DISCOVERY_PANEL.txt',
             'INSTALL_DISK_PANEL.txt',
             'INSTALL_TIMEZONE_PANEL.txt',
             'INSTALL_USERS_PANEL.txt',
             'INSTALL_SUPPORT_PANEL.txt',
             'INSTALL_REVIEW_PANEL.txt',
             'INSTALL_PROGRESS_PANEL.txt',
             'FINISH_PANEL.txt',
             'INSTALL_FAILURE_PANEL.txt']


class HelpDialog(object):
    '''HelpDialog class'''
    def __init__(self, builder):
        '''initialize the helpdialog'''
        self.builder = builder

        self.helpdialog = self.builder.get_object("helpdialog")
        self.textview = self.builder.get_object("helptextview")
        closebutton = self.builder.get_object("helpclosebutton")

        if None in [self.helpdialog, self.textview, closebutton]:
            modal_dialog(_("Internal error"), GLADE_ERROR_MSG)
            raise RuntimeError(GLADE_ERROR_MSG)

        self.helpdialog.connect("delete_event", self.destroy_event)
        if self.helpdialog.window:
            self.helpdialog.window.set_functions(gtk.gdk.FUNC_CLOSE)

        # add bold and underline tags to the text buffer
        text_buffer = self.textview.get_buffer()
        if text_buffer:
            text_buffer.create_tag("bold",
                                   weight=pango.WEIGHT_BOLD)
            text_buffer.create_tag("underline",
                                   underline=pango.UNDERLINE_SINGLE)

        # setup the close button clicked callback
        if closebutton:
            closebutton.connect("clicked", self.closebutton_callback, None)

    def destroy_event(self, widget, event):
        '''catches delete events on the window
        '''
        self.helpdialog.hide_all()

        return True

    def undisplay(self):
        '''convience method to hide the dialog'''
        self.helpdialog.hide_all()

    def display(self, screen):
        '''show the helpdialog and update the help text'''
        self.helpdialog.show_all()

        self.update_help(screen)

    def closebutton_callback(self, widget, data):
        '''Close button callback which hides the dialog'''
        self.helpdialog.hide_all()

    def update_help(self, screen_index):
        '''update the helpdialog text view to match the screen'''
        # ensure that the screen index is not larger then the
        # length of the HELPFILES list
        if screen_index >= len(HELPFILES):
            return

        # get the current locale
        locale_ops = g11nsvc.G11NSvcLocaleOperations()
        locale = locale_ops.getlocale()

        # read the localized help file that matches the current screen
        path = os.path.join(GLADE_DIR, os.path.join('help',
               os.path.join(locale, HELPFILES[screen_index])))
        msg = None
        if os.path.exists(path):
            # the current locale has help text available
            with open(path) as fp:
                msg = fp.read()
        else:
            # try to use the 'C' locale
            path = os.path.join(GLADE_DIR, os.path.join('help',
                   os.path.join('C', HELPFILES[screen_index])))
            if os.path.exists(path):
                # the 'C' locale has help text available
                with open(path) as fp:
                    msg = fp.read()
            else:
                msg = _('Internal Error\n\nNo help available')

        # update the text buffer with the screen's help text
        text_buffer = self.textview.get_buffer()

        # make the first line of text be the title for the help
        first_line_end = msg.find('\n')
        first_line = msg[:first_line_end + 1] + '\n'
        other_lines = msg[first_line_end + 1:]

        # insert the first line of text and make it bold and underlined
        text_buffer.set_text(first_line)
        (text_start, text_end) = text_buffer.get_bounds()
        text_buffer.apply_tag_by_name('bold', text_start, text_end)
        text_buffer.apply_tag_by_name('underline', text_start, text_end)

        # insert the rest of the text
        text_buffer.insert(text_end, other_lines)
